smt/managed/receipt.go

```diff
@@ -222,66 +50,176 @@ func (r *Receipt) String() string {
 	return b.String()
 }
 
+// Validate
+// Take a receipt and validate that the element hash progresses to the
+// Merkle Dag Root hash (MDRoot) in the receipt
+func (r Receipt) Validate() bool {
+	MDRoot := r.Element // To begin with, we start with the object as the MDRoot
+	// Now apply all the path hashes to the MDRoot
+	for _, node := range r.Nodes {
+		// Need a [32]byte to slice
+		hash := node.Hash
+		if node.Right {
+			// If this hash comes from the right, apply it that way
+			MDRoot = MDRoot.Combine(Sha256, hash)
+		} else {
+			// If this hash comes from the left, apply it that way
+			MDRoot = hash.Combine(Sha256, MDRoot)
+		}
+	}
+	// In the end, MDRoot should be the same hash the receipt expects.
+	return MDRoot.Equal(r.MDRoot)
+}
+
+// Copy
+// Make a copy of this receipt
+func (r Receipt) Copy() *Receipt {
+	nr := new(Receipt)
+	nr.Element = append([]byte{}, r.Element...)
+	nr.Anchor = append([]byte{}, r.Anchor...)
+	nr.MDRoot = append([]byte{}, r.MDRoot...)
+	for _, n := range r.Nodes {
+		node := Node{Right: n.Right, Hash: append([]byte{}, n.Hash...)}
+		nr.Nodes = append(nr.Nodes, &node)
+	}
+	return nr
+}
+
+// Combine
+// Take a 2nd receipt, attach it to a root receipt, and return the resulting
+// receipt.  The idea is that if this receipt is anchored into another chain,
+// Then we can create a receipt that proves the element in this receipt all
+// the way down to an anchor in the root receipt.
+// Note that both this receipt and the root receipt are expected to be good.
+func (r *Receipt) Combine(rm *Receipt) (*Receipt, error) {
+	if !bytes.Equal(r.MDRoot, rm.Element) {
+		return nil, fmt.Errorf("receipts cannot be combined. "+
+			"anchor %x doesn't match root merkle tree %x", r.Anchor, rm.Element)
+	}
+	nr := r.Copy()
+	nr.MDRoot = rm.MDRoot
+	for _, n := range rm.Nodes {
+		nr.Nodes = append(nr.Nodes, n)
+	}
```

Comment:
```suggestion
	nr.Nodes = append(nr.Nodes, rm.Nodes...)
```

Reply:
Resolved.  Should have been copying the nodes.